### ISSUE TITLE
Docs for making CreateProcess work on a packaged win32 exe could be improved

### ISSUE THREAD

**Issue by mikehearn** [microsoft/WindowsAppSDK#4651]
### Describe the bug

I'm not sure this is the right repository, but I've seen @DrusTheAxe giving expert answers to related topics here, so I think bugs filed may reach the right people.

There are some aspects of making CreateProcess work correctly in an MSIX packaged app that aren't intuitive and there's a bug in Windows 10 that isn't documented. I had to work it out by trial and error.

On all versions of Windows if your main EXE attempts to execute an EXE that's included in your package as a helper, then it will mysteriously get ERROR_ACCESS_DENIED from CreateProcess ("Access is denied"). Nothing is logged when this happens 😢 

The reason is that *all* EXEs, not just those started from the start menu or a %PATH% execution alias, *must* be listed in the AppX Manifest using `<Application>` elements even if the only thing that starts them is your own code. This isn't mentioned in the docs for `<Application>` nor the docs for `CreateProcess` (in fact the API docs for `CreateProcess` don't contain the word manifest anywhere).

On Win11 it is sufficient to create an entry for your helper EXEs in `AppxManifest.xml`. Unfortunately we must still work on Win10 where it requires more effort:

1. The EXE must have an embedded assembly manifest that contains [the `<msix>` tag](https://learn.microsoft.com/en-us/windows/win32/sbscs/application-manifests#msix).
2. And it must have an `<Application>` element in the AppxManifest
3. *And* (this part is even less documented) it must also have an app execution alias!

In other words every EXE helper tool shipped in an MSIX packaged app must be added to the %PATH% if your runtime uses `CreateProcess` to invoke it, even if you don't actually invoke using that execution alias.

### Steps to reproduce the bug

1. Create an MSIX package with a main EXE (e.g. an Electron app) and a helper EXE.
2. Use `exec` from the `child_process` module to invoke the EXE.
3. Get "Access is denied" on stderr.

### Expected behavior

Everything should just work, obviously. Process creation is the sort of thing you don't expect to break just because of choice of packaging format.

### NuGet package version

None

### Packaging type

Packaged (MSIX)

### Windows version

Windows 11 version 22H2 (22621, 2022 Update), Windows 10 version 22H2 (19045, 2022 Update)

**Comment by DarranRowe** [comment]
One of the big issues I have with this is that the Windows App Runtime provides a way for an application to [request a restart for itself](https://github.com/microsoft/WindowsAppSDK/blob/main/dev/AppLifecycle/AppInstance.cpp#L352). This starts RestartAgent.exe, which is part of the Windows App Runtime package, and isn't mentioned in the AppxManifest file, and exits the process. This [RestartAgent.exe](https://github.com/microsoft/WindowsAppSDK/blob/main/dev/RestartAgent/main.cpp) then executes the process again. This means that not CreateProcess is used twice, but one of them is for a process that is not mentioned in the package's manifest. What's more, there is no alias registered for this, and the restart agent manifest doesn't contain a msix element.
One big thing that I have to ask though, is your package set to run in an appcontainer? The answer to this is yes if the entry point is set to windows.partialTrustApplication, or uap10:TrustLevel is set to appContainer.

**Comment by mikehearn** [comment]
Entry point is "Windows.FullTrustApplication" and `uap10:TrustLevel` doesn't exist in the manifest. Adding the entry points and `<msix>` element as I describe above has resolved the issue for my test Electron app, so I doubt there are any thread attributes that block child processes.

@DrusTheAxe has posted elsewhere on GitHub about how this mechanism works internally, and what he's said does strongly imply that your EXEs have to be listed in the manifest and have the `<msix>` linkage from the assembly manifest on Win10. Reading these scattered posts is how I eventually figured out a combination that works. 

If `RestartAgent.exe` isn't registered and doesn't use `<msix>` in the manifest then maybe it's related to the code you linked to, which does some very complex setup work that nothing else is likely to do, e.g.

https://github.com/microsoft/WindowsAppSDK/blob/main/dev/AppLifecycle/AppInstance.cpp#L419-L426

```
        if (IsPackagedProcess())
        {
            // Desktop Bridge applications by default have their child processes break away from the parent process.  In order to recreate the calling process'
            // environment correctly, this code must prevent child breakaway semantics when calling the agent.  Additionally the agent must do the same when
            // restarting the caller.
            DWORD policy = PROCESS_CREATION_DESKTOP_APP_BREAKAWAY_OVERRIDE;
            THROW_IF_WIN32_BOOL_FALSE(UpdateProcThreadAttribute(attributeList.get(), 0, PROC_THREAD_ATTRIBUTE_DESKTOP_APP_POLICY, &policy, sizeof(policy), nullptr, nullptr));
        }
```

This sounds highly related.

**Comment by mikehearn** [comment]
BTW, I suspect the reason this may be required for Node is this code:

https://github.com/libuv/libuv/blob/5cc7175514571e41cc31219df1ae9dbd8e6b69f9/src/win/process.c#L1072

But really what we need here is someone with source code access to explain or better update CreateProcess docs with an explanation of what combinations yield what outcomes.

**Comment by DarranRowe** [comment]
The use of the extended startup info is documented to keep the package identity. But the implication of the documentation is that  process creation outside of the package should work. Otherwise, PROCESS_CREATION_DESKTOP_APP_BREAKAWAY_ENABLE_PROCESS_TREE being the default doesn't make sense. But yes, not needing the msix element in the application schema is due to the extended startup information. This is documented.
My biggest issue here is that I really don't think CreateProcess should be failing here if the package environment isn't inherited. I could be totally wrong, which is why I am in the process of writing a sample.
Just in case you are wondering where all of this is documented, it is in [UpdateProcThreadAttribute](https://learn.microsoft.com/en-us/windows/win32/api/processthreadsapi/nf-processthreadsapi-updateprocthreadattribute), which is the primary function for setting information in the extended startup information.

**Comment by mikehearn** [comment]
> As another question, if you try to use CreateProcess to launch an executable that isn't in your application's package directory, does it work correctly without any extra information?

Yes this isn't a problem. Only running EXEs inside the app package.

> With regards to Windows 10, I don't think it needs the msix element and an Application entry, but I think it does require some form of telling it to inherit the package identity. 

Sure, but changing the code that runs CreateProcess is hard. It sits inside a bunch of language runtimes that people upgrade slowly, and where code changes are reviewed carefully. Given the docs on this are scattered and hard to interpret, it's hard to convince people to change their code when it works fine outside of the MSIX container. So changing the manifests has to be the way to go here. Clearer direction from Microsoft would help here: CreateProcess could/should discuss all these topics, as package identity is clearly an aspect of the Windows API that becomes increasingly important.

**Comment by mikehearn** [comment]
Continuing my mission to document the hidden internals of CreateProcess, there is ... 🥁  ... a cache! At least on Win10.

1. Create an MSIX with a secondary EXE that lacks the `<msix>` element but is registered in the AppxManifest.
2. Try CreateProcess on it. You get ERROR_ACCESS_DENIED.
3. Add the `<msix>` element in a new version of the MSIX, upgrade.
4. Try CreateProcess on it. It works.
5. Remove the `<msix>` element in a third version of the MSIX, upgrade.
6. Try CreateProcess on it. It still works.

Even rebooting doesn't seem to clean this cache. Changing the MSIX package name forces a miss and restores the "expected" behavior. 

This is quite a common experience when working with this part of Windows. It'd be great if the existence of caches could be reliably documented in the API docs when there's any way to observe the existence of it at all (which there usually is), because it can be extremely confusing otherwise. You make a change that you think should have an effect, it doesn't, that leads to a wild goose chase where you start to wonder if your original understanding was right at all.

**Comment by DrusTheAxe** [comment]
>What's really going on is something like: apps need a global app alias for internal EXEs

What are you trying to do?

* `CreateProcess(pkgdir\some_random.exe)`
* `CreateProcess(pkgdir\app_executable_per_appxmanifest.exe)`
* `CreatePrcess(some_appexecutionalias.exe)`
* `ShellExecute(some_appexecutionalias.exe)`
* ?

**Comment by mikehearn** [comment]
It's the first. `CreateProcess(pkgdir\some_random.exe)`, except:

1. Adding it to the manifest fixes Win11
2. Adding an execution alias fixes it on Win10 even if you don't then use the alias, still `CreateProcess(pkgdir\some_random.exe)`.

**Comment by mikehearn** [comment]
Update: it turns out this is more than a documentation problem. Adding the manifest entries for the helper EXE breaks Windows Store submission because they consider an app that contains **any** CLI exe to be a "headless" app, even if the package also contains a desktop/start menu entry point. And then for some inexplicable reason they require you to apply for special permission to create such apps. Windows happily installs them outside of the store without special permissions, but in store, you have to ask.

So the workaround for this Windows bug breaks Store submission. Great. :( We really do need some sort of fix here. This is a sharp edge that surfaces to most developers as "I packaged with MSIX and now stuff randomly doesn't work".

**Comment by mikehearn** [comment]
@codendone See last comment, I think this is not (only) a documentation bug but a more general problem with Windows.

**Comment by sylveon** [comment]
The issue you're encountering is because you're using `exec`, which spawns a shell (`cmd.exe`) to run your command. Since `cmd.exe` is outside your package (it's in `C:\Windows\System32`), it breaks away, and then `cmd.exe` fails running your executable because it has no package identity and therefore no permission to run the executable. `spawn` works fine here, as long as you don't use `{ shell: true }`, because it directly spawns the exe from within node, instead of going through `cmd.exe`

**Comment by sylveon** [comment]
The easiest solution, if you need to use `exec`, or have a batch file, or can't change the source code, would probably be to [make a PSF fixup](https://learn.microsoft.com/en-us/windows/msix/psf/create-package-support-framework) which hooks `CreateProcess` and adds the proc thread attribute list.

**Comment by sylveon** [comment]
This exists now: https://github.com/sylveon/CreateProcessFixup

**Comment by DrusTheAxe** [comment]
+@dhoehna

**Comment by DrusTheAxe** [comment]
@RDMacLachlan

**Comment by mikehearn** [comment]
Thanks! 

Is it safe to always set `PROCESS_CREATION_DESKTOP_APP_BREAKAWAY_OVERRIDE` like that? That could break apps that do expect to start programs outside their container. Maybe it's safer to restrict the fixup rule to `cmd.exe` only?

**Comment by sylveon** [comment]
It's definitely better to restrict it to only when needed. In my use case however, always setting it worked so I didn't invest time into adding configuration support.

**Comment by mikehearn** [comment]
One thing that's not quite clear to me is why executables that lack package identity can't invoke executables in packages. Shouldn't the OS just set the package identity automatically in that case?

**Comment by sylveon** [comment]
It does work when the executable that's being launched is in the manifest. I suppose when the executable isn't in the manifest, information is missing to grant full package identity.

**Comment by sylveon** [comment]
Turns out, this fixup isn't even needed! There's an (poorly named) `inPackageContext` config for PSF that does exactly what's desired here.
